Dependencies/xr/Source/ARKit/XR.mm

```diff
@@ -0,0 +1,782 @@
+    {
+        // If our camera UVs updated, then also update the FoV to match the updated UVs.
+        [self updateFoV:frame.camera];
+    }
+    
+    // Finally update the XR pose based on the current transform from ARKit.
+    [self updateDisplayOrientedPose:(frame.camera)];
+}
+
+/**
+ Copies the camera buffer from from the current ARFrame into our pixel buffer used for composing the metal texture.
+*/
+- (void)updateFrameBuffer:(CVPixelBufferRef)arCameraBuffer
+{
+    // Lock the frame buffer with read access to stop ARKit from updating the camera buffer during update.
+    CVReturn ret = CVPixelBufferLockBaseAddress(arCameraBuffer, kCVPixelBufferLock_ReadOnly);
+    if (ret != kCVReturnSuccess)
+    {
+        return;
+    }
+    
+    @try
+    {
+        // Find the width, height, and format of the AR image.
+        size_t bufferWidth = CVPixelBufferGetWidth(arCameraBuffer);
+        size_t bufferHeight = CVPixelBufferGetHeight(arCameraBuffer);
+        auto format = CVPixelBufferGetPixelFormatType(arCameraBuffer);
+        
+        // If we don't have two planes the arCamera has not yet been initialized so back out for now.
+        if (CVPixelBufferGetPlaneCount(arCameraBuffer) < 2)
+        {
+            CVPixelBufferUnlockBaseAddress(arCameraBuffer, kCVPixelBufferLock_ReadOnly);
+            return;
+        }
+
+        // Check if the size of the frame buffer has changed, if so dispose of the old one and create a new buffer.
+        if (frameBuffer == nil || bufferWidth != CVPixelBufferGetWidth(frameBuffer) || bufferHeight !=CVPixelBufferGetHeight(frameBuffer))
+        {
+            if (frameBuffer != nil)
+            {
+                CVPixelBufferRelease(frameBuffer);
+                frameBuffer = nil;
+            }
+            
+            // We must specify PixelBufferMetalCompatibility on the frame buffer to make this composable into a Metal Texture.
+            auto attributes = [NSDictionary dictionaryWithObjectsAndKeys:[NSNumber numberWithBool:YES], kCVPixelBufferMetalCompatibilityKey, nil];
+            
+            // Create the frame buffer.
+            ret = CVPixelBufferCreate(kCFAllocatorDefault, bufferWidth, bufferHeight, format, (__bridge CFDictionaryRef)attributes, &frameBuffer);
+            if (ret != kCVReturnSuccess)
+            {
+                throw std::runtime_error("Failed to allocate frame buffer.");
+            }
+        }
+        
+        // Lock the pixel buffer for write access.
+        ret = CVPixelBufferLockBaseAddress(frameBuffer, 0);
+        if (ret != kCVReturnSuccess)
+        {
+            return;
+        }
+        
+        @try
+        {
+            // Copy both planes of the AR camera image to the frame buffer.
+            void* ydestPlane = CVPixelBufferGetBaseAddressOfPlane(frameBuffer, 0);
+            void* ysrcPlane = CVPixelBufferGetBaseAddressOfPlane(arCameraBuffer, 0);
+            memcpy(ydestPlane, ysrcPlane, bufferWidth * bufferHeight);
```

Comment:
Is this what other ARKit examples do? This looks like it is copying pixels out of GPU memory into main memory and then back to GPU memory.

Reply:
We have to do this unfortunately because if you don't then it holds the memory in the ARFrame indefinitely, and the ARKit session freezes after 12 frames.  It might be a side effect of how we are managing the command buffer from the non-UI thread.

I'll give it another look to see if some of the other changes I have made make this no longer necessary.

Reply:
Looks like this is no longer needed! I'm guessing the important change was adding the @autoreleasepool in DrawFrame.